### PR TITLE
Fixed #30427, Fixed #16176 -- Corrected setting descriptor in Field.contribute_to_class().

### PR DESCRIPTION
**Pull request by carltongibson**
Replaces #11337. 

Updates, and puts the tests in place that were hanging from the previous PR. It's difficult to follow so I've commented hoping to make them clear (for now and the next poor soul coming back to this 😀)

We had release notes in an earlier version 5fc4884ba104c502e9aac9b72c2dfec1d7df1560 — those were incorrect since the multiple inheritance example **never** worked. See 4bbe8261c402694a1da3efcaafe3332f9c57af15.

The **change** in behaviour proposed is allowing that structure. See the adjustment in the (now named) `test_multiple_inheritance_allows_inherited_field` method. 

Inheritance works ≈as expected **except** for the _depth-first_ resolution in diamond-shaped cases. Calling that out in `docs/topics/db/models.txt` is I think probably worth while. I need to think about exactly where to put that/what to say. Input welcome. 😜

### CODE REVIEW

File: tests/invalid_models_tests/test_models.py [hunk 1]
```diff
@@ -1212,9 +1212,8 @@ def test_property_and_related_field_accessor_clash(self):
         class Model(models.Model):
             fk = models.ForeignKey('self', models.CASCADE)
 
-            @property
-            def fk_id(self):
-                pass
+        # Override related field accessor.
+        Model.fk_id = property(lambda self: 'ERROR')
```
Comment (adamchainz) [inline tests/invalid_models_tests/test_models.py:1216]: if you can't make this clash via `@property` any more, is the check still valid? Or at least could the error message and/or test model be changed?

File: tests/model_inheritance/tests.py [hunk 2]
```diff
@@ -222,6 +223,36 @@ def test_queryset_class_getitem(self):
...
         self.assertIs(models.QuerySet[Post, int, str], models.QuerySet)
 
+    def test_shadow_parent_attribute_with_field(self):
+        class ScalerParent(models.Model):
+            foo = 1
+
+        class ScalerOverride(ScalerParent):
+            foo = models.IntegerField()
+
+        self.assertEqual(type(ScalerOverride.foo), DeferredAttribute)
```
Comment (adamchainz) [inline tests/model_inheritance/tests.py:233]:
```suggestion
        class ScalarParent(models.Model):
            foo = 1

        class ScalarOverride(ScalarParent):
            foo = models.IntegerField()

        self.assertEqual(type(ScalarOverride.foo), DeferredAttribute)
```

**Comment by adamchainz** [comment]
> Calling that out in docs/topics/db/models.txt is I think probably worth while. I need to think about exactly where to put that/what to say.

Yes definitely. Perhaps just a really terse description is enough...

```
Fields attach descriptors to every model class that includes them.
Because of this, diamond inheritance will cause a leaf model to inherit fields depth-first rather than breadth-first.
```

For anyone attempting diamond inheritance (certainly a small minority) this note may be enough (?). I don't think a long section with diagrams is warranted, given the rarity.

File: tests/model_inheritance/tests.py (hunk 2 above)
Comment (carltongibson) [reply tests/model_inheritance/tests.py:233]: That's Bristol that is.

**Comment by carltongibson** [comment]
I added a small note to the docs.

File: tests/invalid_models_tests/test_models.py (hunk 1 above)
Comment (carltongibson) [reply tests/invalid_models_tests/test_models.py]: I've pushed a second commit removing this test. 

Not 100% convinced... 

We (@felixxm) discussed previously on the original PR. IIRC we said we'd leave it, and maybe adjust to add a check going the other way. (i.e. Cover _I tried to define the `pk_id` property, but it got overridden_ vs the previous _Because I'd defined the property the field descriptor wasn't set_.) 

I'd like to address that as a separate cleanup ticket. 

* This has been going on too long already. 
* And there are real issues being fixed here vs a system check/config error. 

🤔

File: tests/model_inheritance/test_abstract_inheritance.py [hunk 3]
```diff
@@ -50,14 +55,76 @@ class Meta:
...
+        """
+        In contrast to standard Python MRO, resolution of inherited fields is
+        depth-first, rather than breadth-first.
+
+        This is because a copy of the parent field descriptor is placed onto
+        the model class in ModelBase.__new__(), rather than the attribute
+        lookup going via bases. (It only **looks** like inheritance.)
+
+        Here, Child inherits name from Root, rather than ParentB.
+        """
```
Comment (jarekwg) [inline tests/model_inheritance/test_abstract_inheritance.py:73]: Would it be worth emphasising that this straying from pythonic expectation is an _undesirable_ behaviour?
In case `ModelBase.__new__()` were to get rewritten in future, and whoever works on it ends up thinking they need to be supporting these testcases?
Orr just writing the testcase as an expected failure would imply this.

Comment (carltongibson) [reply tests/model_inheritance/test_abstract_inheritance.py:73]: I don't see it as undesirable. Just a fact about the way models are built. 

We already document that inheritance isn't quite like Python inheritance. This is just part of that. 

It's not likely that rewriting `ModelBase.__new__()` is really on the cards.

File: docs/topics/db/models.txt [hunk 4]
```diff
@@ -1457,6 +1457,13 @@ different database tables).
 Django will raise a :exc:`~django.core.exceptions.FieldError` if you override
 any model field in any ancestor model.
 
+Note that because of the way fields are resolved during class definition, model
+fields inherited from multiple abstract parent models are resolved in a strict
+depth-first order. This contrasts with standard Python MRO, which is resolved
+breadth-first in cases of diamond shaped inheritance. This difference only
+affects complex model hierarchies, which (as per the advice above) you should
+try to avoid.
```
Comment (adamchainz) [inline docs/topics/db/models.txt:1465]: Nice, I like the admonition.

File: tests/invalid_models_tests/test_models.py (hunk 1 above)
Comment (adamchainz) [reply tests/invalid_models_tests/test_models.py]: Okay fine by me, as long as the ticket exists :)

**Review by adamchainz — APPROVED** [review]

File: tests/model_inheritance/tests.py (hunk 2 above)
Comment (adamchainz) [reply tests/model_inheritance/tests.py]: I don't know what you're saying but it sounds rude in cockney rhyming slang... https://en.wiktionary.org/wiki/bristol#Noun

File: tests/model_inheritance/test_abstract_inheritance.py [hunk 5]
```diff
@@ -50,14 +55,76 @@ class Meta:
...
+
+            class Meta:
+                abstract = True
+
+        class Child(ParentA, ParentB):
+            name = models.IntegerField()
+
+        self.assertEqual(Child.check(), [])
+        inherited_field = Child._meta.get_field('name')
+        self.assertTrue(isinstance(inherited_field, models.IntegerField))
```
Comment (felixxm) [inline tests/model_inheritance/test_abstract_inheritance.py:127]: `test_target_field_may_be_pushed_down` works without a patch. I would move it to a separate commit to make it clear where the behavior has changed.

**Review by felixxm — APPROVED** [review]
@carltongibson Thanks for pushing this to the finish line :checkered_flag: 

@jarekwg Thanks for all your efforts  :1st_place_medal: 

@adamchainz Thanks for reviews :detective:

**Comment by carltongibson** [comment]
Thanks both. I'll tidy this up and pull it in. 

(What a set of tickets 😅)

File: tests/invalid_models_tests/test_models.py (hunk 1 above)
Comment (carltongibson) [reply tests/invalid_models_tests/test_models.py:1216]: Create follow-up at ticket-32847

(Will leave this conversation unresolved so folks maybe see it as they're browsing.)

**Comment by randykleinman** [comment]
Will this be backported to 3.2.x?

**Comment by adamchainz** [comment]
No, the backport policy only allows “Security fixes and data loss bugs” for a feature change like this: https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions .

**Comment by randykleinman** [comment]
Ok, thanks!
